clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -183,4 +209,24 @@ public synchronized long lastSuccessfulUpdate() {
     public long refreshBackoff() {
         return refreshBackoffMs;
     }
+
+    /**
+     * Get the preferred node for getting metadata update from (if there is one)
+     * @return The preferred node or null if there is none
+     */
+    public synchronized Node updateFrom() {
+        return updateFrom;
+    }
+
+    /**
+     * Add a listener to get notifications when metadata is updated.
+     * @param listener The listener
+     */
+    public synchronized void addListener(MetadataListener listener) {
+        this.listeners.add(listener);
+    }
+
+    public interface MetadataListener {
```

Comment:
Nit, but since this is a nested class that will probably always be referred to by `Metadata.MetadataListener`, `Listener` would probably be more concise/less redundant.